### ISSUE TITLE
Firefox crashes if pipewire-screenaudio is active on NixOS with Mesa GPU drivers

### ISSUE THREAD

**Issue by flippette** [IceDBorn/pipewire-screenaudio#102]
I've tested both Discord and the [gUM Test Page](https://mozilla.github.io/webrtc-landing/gum_test.html), both exhibits this behavior.

I'm using Firefox v127.0.1 on NixOS 24.11 unstable.

(Edit: More system info:

- OS: \<as above\>
- Firefox: \<as above\>
- WM: Hyprland v0.41.1
- libpipewire: v1.0.7

I'll attach more info on request.)

**Comment by IceDBorn** [comment]
I have this issue too. @jim3692 does not have this bug on Arch, I'm not sure if he's on the latest stable firefox though.

**Comment by IceDBorn** [comment]
We need more testing from NixOS users as well as different distro users. We're not sure if this issue is NixOS specific or not.

**Comment by IceDBorn** [comment]
We did some debugging, and it seems like that this is a bug that occurred with Firefox 127.0 specifically on nix.

**Comment by IceDBorn** [comment]
`[Parent 170908, IPC I/O Parent] WARNING: process 171266 exited on signal 11: file /build/firefox-128.0/ipc/chromium/src/base/process_util_posix.cc:335`, this error is thrown in the shell.

**Comment by jim3692** [comment]
It doesn't crash if I set the `latency` to 17 or more. Given that my display is 60Hz (meaning the frametime is 16.67ms), there must be a relation to those numbers.

https://github.com/IceDBorn/pipewire-screenaudio/blob/bd404d0ee4b32fb797f2dcdeee3b6e773c06455c/extension/scripts/index.js#L44-L47

**Comment by flippette** [comment]
> @flippette Can you tinker with the latency and see if it fixes the crashes for you?

My Nix install broke, so I'm back on Arch for the time being. I'll try doing that when I get my laptop.

**Comment by jim3692** [comment]
@flippette Could you tell us what is your GPU? I am starting to suspect the AMD drivers, as Firefox isn't crashing in VirtualBox.

**Comment by flippette** [comment]
> @flippette Could you tell us what is your GPU? I am starting to suspect the AMD drivers, as Firefox isn't crashing in VirtualBox.

I have a Radeon RX 7800 XT.

**Comment by jim3692** [comment]
Thank you for your feedback. I tried to passthrough my RTX 2060 to a fresh NixOS VM. Firefox doesn't crash with nvidia either. I hate this bug.

**Comment by IceDBorn** [comment]
So, it's a NixOS + RADV or AMDVLK specific issue.

**Comment by flippette** [comment]
I am using RADV, not sure about AMDVLK.

**Comment by IceDBorn** [comment]
![signal-2024-07-24-201245_002](https://github.com/user-attachments/assets/cab40e88-7283-4518-99ba-26a1da071bfa)

This setup does not crash the tab. I found this by changing the injector script to connect "Noise canceling source" instead of "pipewire-screenaudio", and the stream successfully started, having audio from my microphone. Doing this resulted in the node with only one capture port getting connected to Librewolf. So, the solution would be to replicate what "Noise canceling source" does.

**Comment by jim3692** [comment]
Hello all,

I got it crashing on my older PCs with NixOS Unstable:
- AMD RX550
- Intel HD 4600 (after [enabling VA-API](https://nixos.wiki/wiki/Accelerated_Video_Playback))

**Comment by IceDBorn** [comment]
We have to try using NVK instead of the proprietary drivers and see if it crashes for nvidia too @jim3692.

**Comment by jim3692** [comment]
> We have to try using NVK instead of the proprietary drivers and see if it crashes for nvidia too @jim3692.

Firefox is also crashing with my RTX 2060 with the NVK driver, even though Firefox is not reporting hardware decoding in the `about:support` page

**Comment by IceDBorn** [comment]
@flippette it works for me now on the latest unstable, can you confirm?

**Comment by flippette** [comment]
I'm still on Arch for the time being, so I can't say. Sorry 😅

**Comment by IceDBorn** [comment]
@flippette you betrayed the nix cult...

**Comment by DADA30000** [comment]
there is also an active firefox issue specific to nix, using firefox-bin fixes discord tab crashing, but idk how to use pipewire-screenaudio with firefox-bin

**Comment by IceDBorn** [comment]
I haven't tested, but, wouldn't installing the firefox package as shown at the [example](https://github.com/IceDBorn/pipewire-screenaudio/blob/bd404d0ee4b32fb797f2dcdeee3b6e773c06455c/README.md?plain=1#L41) work?

**Comment by DADA30000** [comment]
> I haven't tested, but, wouldn't installing the firefox package as shown at the [example](https://github.com/IceDBorn/pipewire-screenaudio/blob/bd404d0ee4b32fb797f2dcdeee3b6e773c06455c/README.md?plain=1#L41) work?

nope, doesn't work like that with firefox-bin unfortunately, just says "native connector is misconfigured"

**Comment by IceDBorn** [comment]
@DADA30000 Does it still crash for you? Because I'm pretty sure that it does not for me. I'll try re-testing tomorrow.

**Comment by DADA30000** [comment]
> @DADA30000 Does it still crash for you? Because I'm pretty sure that it does not for me. I'll try re-testing tomorrow.

Just updated my flake, rebuilded my system, it still does, it crashes only when you're in a call with someone, and that someone says something, or when you go into voice settings in discord, and do a microphone check

**Comment by IceDBorn** [comment]
@DADA30000 I'll test it out too then.

**Comment by DADA30000** [comment]
> @DADA30000 I'll test it out too then.

could you maybe do something so that it would work with bin version? Because I think that this issue with firefox on NixOS will take a lot of time to be fixed :D

**Comment by IceDBorn** [comment]
I'm not sure what the steps are to get it working with the bin package. Maybe you could follow the install instructions for a normal FHS distro. Ex. add the manifest.json on ~/.mozilla/native-messaging-hosts. https://github.com/IceDBorn/pipewire-screenaudio/blob/main/install.sh. You'll just have to change the path to the store path of pipewire-screenaudio.

**Comment by DADA30000** [comment]
> I'm not aware of the steps needed to get it working with the bin package. Maybe you could follow the install instructions for a normal FHS distro. Ex. add the manifest.json on ~/.mozilla/native-messaging-hosts. https://github.com/IceDBorn/pipewire-screenaudio/blob/main/install.sh. You'll just have to change the path of pipewire-screenaudio to the nix store one.

hmm, ok, I will try that, thanks!

**Comment by DADA30000** [comment]
> I'm not aware of the steps needed to get it working with the bin package. Maybe you could follow the install instructions for a normal FHS distro. Ex. add the manifest.json on ~/.mozilla/native-messaging-hosts. https://github.com/IceDBorn/pipewire-screenaudio/blob/main/install.sh. You'll just have to change the path of pipewire-screenaudio to the nix store one.

that worked! Idk how I haven't thought of that, thanks again!

**Comment by IceDBorn** [comment]
So, issue's fixed using the bin I guess. We just have to add a home manager module.

**Comment by DADA30000** [comment]
> So, issue's fixed using the bin I guess. We just have to add a home manager module.

oh wait, there is another problem, the only option that I can select in audio sources are virtual surround sink (it's not my main output) and audio from various tabs with audio playing

**Comment by DADA30000** [comment]
Oh, sorry. it's the same on non-bin version, unrelated to this issue

**Comment by DADA30000** [comment]
@IceDBorn you know how to fix a problem where 'All Desktop Audio" is not listed?

**Comment by IceDBorn** [comment]
All desktop audio is currently broken. Will have to push a fix.

**Comment by IceDBorn** [comment]
This seems fixed on the latest unstable of nixpkgs. Closing...
